### PR TITLE
Support for Go Programs

### PR DESCRIPTION
**Pull request by ArquintL**
This PR mainly enables Gobra to verify `.go` files with annotations in comments, i.e. `//@` and `/*@ ... @*/`.
However, several other features are contained in this PR:
- in-memory caching of `PProgram` meaning that the parser does not need to be invoked on files that have not changed (only useful when using Gobra together with Gobra Server). Caching is turned off when using Gobra via CLI
- `--module` flag to pass a prefix that should be ignored when resolving import paths
- Equality on slices (thanks to @Felalolf)
- Predicate instances can be used as triggers
- domains are treated as ghost meaning that e.g. mathematical types are allowed
- arbitrary import paths (without whitespaces) are now accepted
- error locations in CLI include filename 
- a few basic unit tests have been added for the process of gobrafying a `.go` file
- [CI] `viper-toolchain-versions.sh` centrally stores the versions of our Viper dependencies such that they are no longer duplicated in several workflow files and dockerfiles

Open tasks
- [ ] Decide on the reporting of parse & type errors from imported packages & implement it (@ArquintL)
- [ ] Finalize encoding of slices (@Felalolf)

### CODE REVIEW

File: viper-toolchain-versions.sh [hunk 1]
```diff
@@ -0,0 +1,5 @@
+#!/bin/sh
+
+SILVER_REF="f645539a04e76d70337d7b5b108e7f9a9de9bc7a"
```
Comment (ArquintL) [inline viper-toolchain-versions.sh:3]: This corresponds to [this commit](https://github.com/viperproject/silver/commit/f645539a04e76d70337d7b5b108e7f9a9de9bc7a). We can no longer use the Viper release as `AbstractGobraTests` needs to specialize what a comment is and what not such that the annotation matching correctly works for `.go` test cases

File: src/main/scala/viper/gobra/frontend/Parser.scala [hunk 2]
```diff
@@ -86,8 +95,37 @@ object Parser {
       }
     }
 
+    def parseSourceCached(source: FromFileSource): Either[Vector[VerifierError], PProgram] = {
+      var cacheHit = true
+      def parseAndStore(): (Either[Vector[VerifierError], PProgram], Positions) = {
+        cacheHit = false
+        val res = parseSource(source)
+        sourceCache += (source.path.toString, source.content) -> (res, positions)
```
Comment (jcp19) [inline src/main/scala/viper/gobra/frontend/Parser.scala:103]: Having the entire content of the program in the Map key might cause unnecessary slowdowns in the cache lookup and may cause unnecessary memory consumption (for big projects). Could we instead store an hash of the AST?

**Review by jcp19 — COMMENTED** [review]
I did not look into the slice encoding because that is not yet finished. In general, it looks fine but I have some comments and questions. Also, it seems that we only have two tests for `.go` files. I believe that this is working well for `.go` files because you have used this feature in the case study, but it would be nice to have more tests for feature reference.

File: viper-toolchain-versions.sh [hunk 3]
```diff
@@ -0,0 +1,5 @@
...
+SILICON_REF="v.21.07-release"
+CARBON_REF="v.21.07-release"
```
[the remark this answers is not in the document]
Comment (ArquintL) [reply viper-toolchain-versions.sh:5]: done

File: src/main/scala/viper/gobra/frontend/Config.scala [hunk 4]
```diff
@@ -105,6 +110,12 @@ class ScallopGobraConfig(arguments: Seq[String], isInputOptional: Boolean = fals
     descr = "List of Gobra programs or a single package name to verify"
   )
 
+  val module: ScallopOption[String] = opt[String](
+    name = "module",
+    descr = "Name of current module that should be used for resolving imports",
+    default = Some("")
+  )
```
[the remark this answers is not in the document]
Comment (ArquintL) [reply src/main/scala/viper/gobra/frontend/Config.scala:117]: Exactly, it's the module name you typically write into the `go.mod` file. We can think about integrating a `go.mod` into Gobra

File: src/main/scala/viper/gobra/frontend/Parser.scala (hunk 2 above)
Comment (ArquintL) [reply src/main/scala/viper/gobra/frontend/Parser.scala:103]: I'm taking now the MD5 hash of the path and content as the key

File: src/main/scala/viper/gobra/frontend/Gobrafier.scala [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+  private val args = "(.*?)"
+  private val argList = "\\((.*?)\\)"
+  private val functionInvocation = s"$identifier$argList"
+  private val functionDecl = s"$func_keyword\\s+$identifier\\s*$argList\\s*($argList)?"
+  private val spec = "(.*?)"
+  private val vars = "(.*?)"
+  private val predicate = "(.*?)"
+  private val assignment = s"$vars\\s*(:=|=)\\s*$vars"
+  private val goifiedComment = s"([;]\\s*(.*?))?"
+
```
[the remark this answers is not in the document]
Comment (ArquintL) [reply src/main/scala/viper/gobra/frontend/Gobrafier.scala:42]: it is indeed not there. However, you can always simply place it in a spec comment, i.e.
```
/*@
type yourDomain domain {

}
@*/
```

File: src/main/scala/viper/gobra/frontend/Gobrafier.scala [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+      addressableVariables.foreach(addrVar => {
+        // remove all control characters
+        val variable = addrVar.filter(_ >= ' ')
+
+        code = code.replaceAll(variable, variable + "!")
+      })
+
+      code
+    })
```
[the remark this answers is not in the document]
Comment (ArquintL) [reply src/main/scala/viper/gobra/frontend/Gobrafier.scala:179]: great catch!

File: src/main/scala/viper/gobra/frontend/Gobrafier.scala [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+      "\n" + // added to maintain line consistency
+      m.group(2) +
+      "func " +
+      m.group(3) +
+      parens(m.group(4) +
+      (if (m.group(4) == "") "" else ", ") +
+      addGhostKeywordToParamsList(m.group(1))) +
+      " " +
+        (if (m.group(5) == null) "" else m.group(5))
+    })
```
[the remark this answers is not in the document]
Comment (ArquintL) [reply src/main/scala/viper/gobra/frontend/Gobrafier.scala]: I've now added a few more basic test cases to cover at least every regex we have (and found & fixes a couple of bugs while doing so ^^)

File: src/main/scala/viper/gobra/frontend/Gobrafier.scala [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+      "\n" + // added to maintain line consistency
+      functionSpec +
+      "func " +
+      functionName +
+      parens(actualParams +
+      (if (actualParams == "") "" else ", ") +
+      addGhostKeywordToParamsList(ghostParamsList)) +
+      " " +
+        (if (returns == null) "" else returns)
+    })
```
Comment (jcp19) [inline src/main/scala/viper/gobra/frontend/Gobrafier.scala:121]: Thanks for making this so much easier to read :D

**Review by jcp19 — APPROVED** [review]
Looks good!

**Pull request merged**